### PR TITLE
Do not require preprocessing.py to redundantly load datasets it just created

### PR DESCRIPTION
**Pull request by bpopeters**
This pull request is the first in a series designed to clarify the preprocessing code so that it is easier to maintain and extend. It does this by changing some misleading function names, changing out-of-date documentation, and changing the structure of preprocess.py so that 

1. Building objects and serializing them are not handled in the same function
2. It is not necessary to reload a serialized dataset immediately after it was created (this was not obvious because it happened under a couple layers of method calls)

### CODE REVIEW

**Comment by bpopeters** [comment]
@srush I looked at the logs for the failing tests and I don't understand what's going on with them. They don't seem to have anything to do with the changes I made.

**Comment by vince62s** [comment]
@bpopeters are you sure that in shard mode it will be ok in terms of Ram ? I read your comment aboit memory reclaim by the OS, did you actually test with a large dataset ?

**Comment by bpopeters** [comment]
I haven't tested it, but it's also not my claim. I was just clearing up the comment @JianyuZhan wrote. My changes should not alter the sharding behavior.

**Comment by vince62s** [comment]
oh ok sorry. I thought it was loading all shards in memory since it was not unloading them. never mind.

**Comment by bpopeters** [comment]
Update: this last commit changes the implementation of `get_fields` so that it no longer redirects traffic to various static methods defined on dataset subclasses. I have a few reasons for doing this:

1. The code defined on these dataset subclasses is largely redundant. The same code for making target text fields is written in three different places. This clutters up the code base and makes it easier for bugs to enter the code and more difficult for them to be detected. For example, the `src_map` and `alignment` fields are always defined in preprocessing, even though they will disregarded unless you're using copy attention. It was difficult to see this from the previous implementation of `get_fields`.
2. Static methods are usually undesirable because they muddle up the purpose of classes. The purpose of a dataset should be as a container for training examples, not as a toolbox in which you define a bunch of utility functions.
3. In general, I think we should work toward separating fields from datasets. The purpose of a field is to define how raw data can be processed into tensors. The purpose of a dataset is to contain the examples that are created from applying fields to raw data. An ideal dataset should not have to care about what kind of fields it uses: it just needs to be able to apply those fields to produce parallel data. Right now we have a variety of dataset subclasses, each of which is designed to support a single pair of field types (at the moment, the supported ones are (text, text), (audio, text), and (image, text)). This is undesirable because you need to define a new dataset class every time you want to work with a novel pair of modalities. In the long run, I think we should have a single dataset that can accept an arbitrary pair of fields. Removing the code where the datasets create their fields can be seen as the first step toward that goal.

Bonus: after pushing this commit, the PR no longer fails tests.

**Comment by bpopeters** [comment]
More updates:
1. I changed the implementation if `get_num_features` in inputter.py in an analogous way to the changes to `get_fields` from yesterday.
2. I removed the `_construct_example_fromlist` private method from `DatasetBase` because it is redundant.

**Comment by bpopeters** [comment]
Another update: I changed the implementation of `inputter.build_vocab` so that it takes advantage of the `build_vocab` method of torchtext `Field` objects instead of reimplementing basically the same functionality. This required a slight reordering in preprocessing.py: currently, we remove a `Dataset`'s fields before saving and the `build_vocab` method requires the datasets to still have fields. So now the datasets are saved after the vocab is created, not before. 

However, don't think we need to remove the fields anymore if we update to pytorch 0.4.1. In contrast to previous versions, in 0.4.1 `torch.dtype` can be serialized. The other obstacle to pickling fields is lambda expressions, but a lambda expression can always be replaced by a normal function, which _can_ be pickled. `torchtext.text.Field`s cannot be pickled with default parameters only because the default tokenizer is `lambda s: s.split()`, which can be trivially replaced by `str.split`, for example.

**Comment by bpopeters** [comment]
This is still very much in progress. Besides a few bugs, some of the changes I've made require 0.4.1.

**Comment by bpopeters** [comment]
Mission creep has turned this into a fairly big effort to simplify the preprocessing code. I'm still working on it, but it'll be a while before it's mergeable and I'll close the pull request in the meantime.
